lib/src/render/documentation_renderer.dart

```diff
@@ -20,16 +23,16 @@ class DocumentationRendererHtml implements DocumentationRenderer {
   DocumentationRenderResult render(
```

Comment:
This is a good place for this code.  I had considered inserting this somewhere in the markdown rendering pipeline too to try to check embedded HTML, but perhaps users of the markdown output already have their own sanitization in place.

Reply:
`package:markdown` explicitly says to do sanitization as a post-processing step:
https://pub.dev/packages/markdown#html-sanitization

They also reference [Markdown's XSS Vulnerability (and how to mitigate it)](https://github.com/showdownjs/showdown/wiki/Markdown%27s-XSS-Vulnerability-(and-how-to-mitigate-it)).

TL;DR: XSS mitigation is best done post-markdown rendering.

Maybe it's possible to write a custom implementation of [HtmlRenderer](https://pub.dev/documentation/markdown/latest/markdown/HtmlRenderer-class.html) which does sanitizing on-the-fly. But I'm not sure if `Text` nodes in markdown can contain inline HTML, as is `HtmlRenderer` does not escape `<>` when rendering `Text` nodes.

Reply:
I took a quick stab a it, and it doesn't look easy t make `HtmlRenderer` produce safe HTML... Even just using `htmlSerializeEscape` when outputting `Text` broke a lot of tests that aimed to ensure that inline HTML keeps working.

My two cents is that maybe... but probably we have to sanitize after markdown rendering -- otherwise, we'll probably have to dramatically refactor our markdown rendering pipeline.

-------------------

If we wanted to support `{@tool}` directives which can side-step HTML sanitizing, then I suggest we extend the markdown parser with custom block syntax, and have it translate `{@tool foo --my-flag}...{@end-tool}` into `<pre data-tool="foo --my-flag">...</pre>`, then we could replace these `<pre>` tags after HTML sanitizing, run tools at this point and decide if we want to sanitize the output of a tool on a tool by tool basis.

Played a bit around with `package:markdown`, and we can do things like:
```dart
final _beginToolPattern = RegExp(r'^{@tool ([^}\n]+)}$');
final _endToolPattern = RegExp(r'^{@end-tool}');

class ToolBlockSyntax extends BlockSyntax {
  @override
  RegExp get pattern => _beginToolPattern;

  const ToolBlockSyntax();

  @override
  bool canParse(BlockParser parser) {
    return pattern.hasMatch(parser.current);
  }

  @override
  List<String> parseChildLines(BlockParser parser) {
    var childLines = <String>[];
    parser.advance();

    while (!parser.isDone && !_endToolPattern.hasMatch(parser.current)) {
      childLines.add(parser.current);
      parser.advance();
    }
    if (!parser.isDone) {
      parser.advance();
    }

    return childLines;
  }

  @override
  Node parse(BlockParser parser) {
    var match = pattern.firstMatch(parser.current)!;
    var toolCommand = match.group(1)!;

    var childLines = parseChildLines(parser);
    // We should always end with a line-break (just makes sense, I think)
    childLines.add('');

    var text = childLines.join('\n');
    var pre = Element.text('pre', escapeHtml(text));
    pre.attributes['data-tool'] = escapeHtmlAttribute(toolCommand);

    return pre;
  }
}
```

Then we just have find `<pre data-tool="COMMAND">STDIN</pre>` afterwards, decode the HTML text and attribute text before running the tool. Anyways, maybe that's a larger refactor :D

Reply:
Ugh.   Moving tool execution to the back end rather than the front breaks all kinds of internal assumptions in dartdoc and runs into a sync vs. async problem as well IIRC from the last time I tried to change the pipeline.

When you say that execution of tools "might not need to happen here" do you mean that it literally doesn't need to run, or just that it doesn't have to be moved?  I do wonder if we can get away with "not solving this" for pub for the moment -- if we just forbid tool execution when the post-processing flag is on and then leave me the cleanup to get this working in a more comprehensive manner.   IIRC tool execution is forbidden anyhow on pub?

Reply:
Thanks for the detailed notes on all of this!

Thinking about this from dartdoc's perspective going with \#1 seems reasonable for now.   Being as safe as possible seems a good default choice in any event.  But we're going to have to do something about Flutter as script tags are part of their existing tool infrastructure.  Perhaps we can make the sanitizer able to be disabled in dartdoc_options.yaml so it can be scoped?   Pub already will filter things from dartdoc_options files so that seems like it could be reasonable to do (as long as pub knows that dartdoc_options.yaml files can be located anywhere in the tree?).

Examples from Flutter docs:

Trusted domain iframe from a `youtube` directive:  just an iframe
  https://api.flutter.dev/flutter/material/material-library.html

Here is an example of a tool inserting script tags:
  https://api.flutter.dev/flutter/material/AboutListTile-class.html

Reply:
> Trusted domain iframe from a youtube directive:
> https://api.flutter.dev/flutter/material/material-library.html

I think it would be easy allow iframes from trusted domains. Or we could simply hardcode an exception for youtube, that seems reasonable given how many users want to embed youtube.

But in general, I suspect `<iframe sandbox="allow-same-origin allow-scripts"` is reasonably safe, so long as the embedded document is, either: (i) from a separate domain, or (ii) trusted (if it's from the same domain). It seems like a reasonable expectation that if the document is from the same domain, then it's trusted.

> Here is an example of a tool inserting script tags:
> https://api.flutter.dev/flutter/material/AboutListTile-class.html

Embedding `<script>` inline violates a lot of good practices, include security recommendations around content-security-policies (I think we strive to use something like [strict CSP](https://csp.withgoogle.com/docs/strict-csp.html) whenever possible).

In many cases, it would be better if the `{@tool` directive generated something like:
```
<div data-snippet-buttons="longSnippet1"></div>
```

Then `dartdoc` options to inject a `<script src="static/snippet-buttons.js">` file in the header. Then have `snippet-buttons.js` do something like:
```js
window.addEventListener('DOMContentLoaded', function() {
  document.querySelectorAll('[data-snippet-buttons]').forEach(function(div) {
    // Do whatever you want to do with `div`, setup child elements and onclick handlers...
  });
});
```

Hmm, looking at it Flutter docs already embed `https://api.flutter.dev/assets/snippets.js`, so instead of embedding `onclick` handlers in the HTML code, they just have use `data-...` properties (or special CSS classes) to indicate that an HTML element should `onclick` handlers, and set it up in an `DOMContentLoaded` handler in `snippets.js`.

It's similar to what we do with syntax highlighting, we just apply a special CSS class `language-<language>` on a `<pre>` tag, and then `highlight.pack.js?v1` will do the rest in an `DOMContentLoaded` handler.

--------

So perhaps we should just allow custom `data-` attributes, or we could limit it to only allow data-attributes on the form `data-custom-*` (that way you can't mess with `data-base-href` and `data-using-base-href` which are already used in dartdoc generate output).